src/lib.rs

```diff
@@ -1365,6 +1364,109 @@ impl DnsCache {
     }
 }
 
+pub trait AsAddr {
+    fn as_addr(&self) -> Result<HashSet<Ipv4Addr>>;
+}
+
+impl<T: AsAddr> AsAddr for &T {
+    fn as_addr(&self) -> Result<HashSet<Ipv4Addr>> {
+        (&self).as_addr()
+    }
+}
+
+impl AsAddr for &str {
+    fn as_addr(&self) -> Result<HashSet<Ipv4Addr>> {
+        let res = self
+            .split(',')
+            .map(|addr| (addr, std::net::Ipv4Addr::from_str(addr)))
+            .map(|(raw, addr)| {
+                // i'd really like to avoid the vec![] allocation here
+                match addr {
+                    Ok(ok) => Ok({
+                        let mut h = HashSet::new();
+                        h.insert(Ipv4Addr::from_std(&ok));
+                        h
+                    }),
+                    // on ipnet try to parse it as an CIDR instead as a last ditch effort
+                    #[cfg(feature = "ipnet")]
+                    Err(_) => ipnet::Ipv4Net::from_str(raw)
+                        .map_err(|err| Error::ParseIpAddr(err.to_string()))
+                        .and_then(|net| net.as_addr()),
+                    #[cfg(not(feature = "ipnet"))]
+                    Err(err) => Err(Error::ParseIpAddr(err.to_string())),
+                }
+            })
+            // this collect isn't ideal, but helps us remove the result err
+            .collect::<Result<Vec<_>>>()?
+            .into_iter()
+            .flatten()
+            .collect();
+
+        Ok(res)
+    }
+}
+
+impl AsAddr for String {
+    fn as_addr(&self) -> Result<HashSet<Ipv4Addr>> {
+        self.as_str().as_addr()
+    }
+}
+
+impl<I: AsAddr> AsAddr for &[I] {
+    fn as_addr(&self) -> Result<HashSet<Ipv4Addr>> {
+        let ips = self
+            .into_iter()
+            .map(|val| val.as_addr())
+            .collect::<Result<Vec<HashSet<_>>>>()?
+            .into_iter()
+            .flatten()
+            .collect();
```

Comment:
I would suggest creating a new empty `HashSet` first, then use a simple loop to iterator the `&[T]` and fill in the new set.